// loadmill/lib/index.js
"use strict";
var tslib_1 = require("tslib");
require("./polyfills");
var fs = require("fs");
var superagent = require("superagent");
var utils_1 = require("./utils");
var reporter_1 = require("./reporter");
function Loadmill(options) {
    var _a = options, token = _a.token, _b = _a._testingServerHost, _testingServerHost = _b === void 0 ? utils_1.TESTING_HOST : _b;
    var testingServer = "https://" + _testingServerHost;
    var testSuitesAPI = testingServer + "/api/test-suites";
    var testPlansAPI = testingServer + "/api/test-plans";
    function _runFolderSync(listOfFiles, execFunc) {
        var funcArgs = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            funcArgs[_i - 2] = arguments[_i];
        }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var results, _a, listOfFiles_1, file, res, testResult;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        results = [];
                        _a = 0, listOfFiles_1 = listOfFiles;
                        _b.label = 1;
                    case 1:
                        if (!(_a < listOfFiles_1.length)) return [3 /*break*/, 5];
                        file = listOfFiles_1[_a];
                        return [4 /*yield*/, execFunc.apply(void 0, [file].concat(funcArgs))];
                    case 2:
                        res = _b.sent();
                        return [4 /*yield*/, _wait(res)];
                    case 3:
                        testResult = _b.sent();
                        results.push(testResult);
                        if (!testResult.passed)
                            return [3 /*break*/, 5];
                        _b.label = 4;
                    case 4:
                        _a++;
                        return [3 /*break*/, 1];
                    case 5: return [2 /*return*/, results];
                }
            });
        });
    }
    function _wait(testDefOrId, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var resolve, reject, testDef, apiUrl, webUrl, intervalId;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                testDef = typeof testDefOrId === 'string' ? {
                    id: testDefOrId,
                    type: Loadmill.TYPES.LOAD
                } : testDefOrId;
                apiUrl = getTestAPIUrl(testDef, testingServer);
                webUrl = getTestWebUrl(testDef, testingServer);
                intervalId = setInterval(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                    var body, bodyWithFlows, testResult, err_1;
                    return tslib_1.__generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                _a.trys.push([0, 5, , 6]);
                                return [4 /*yield*/, superagent.get(apiUrl)
                                        .auth(token, '')];
                            case 1:
                                body = (_a.sent()).body;
                                if (!isTestInFinalState(body)) return [3 /*break*/, 4];
                                clearInterval(intervalId);
                                if (!(testDef.type === Loadmill.TYPES.TEST_PLAN)) return [3 /*break*/, 3];
                                return [4 /*yield*/, superagent.get(apiUrl + "?fetchAllFlows=true").auth(token, '')];
                            case 2:
                                bodyWithFlows = (_a.sent()).body;
                                body = bodyWithFlows;
                                _a.label = 3;
                            case 3:
                                testResult = tslib_1.__assign({}, testDef, { url: webUrl, description: body && body.description, passed: isTestPassed(body, testDef.type), startTime: body.startTime, endTime: body.endTime });
                                if (testDef.type === Loadmill.TYPES.SUITE) {
                                    testResult.flowRuns = reductFlowRunsData(body.testSuiteFlowRuns);
                                }
                                else if (testDef.type === Loadmill.TYPES.TEST_PLAN) {
                                    testResult.testSuitesRuns = reductTestSuitesRuns(body.testSuitesRuns, testingServer);
                                }
                                if (callback) {
                                    callback(null, testResult);
                                }
                                else {
                                    resolve(testResult);
                                }
                                _a.label = 4;
                            case 4: return [3 /*break*/, 6];
                            case 5:
                                err_1 = _a.sent();
                                clearInterval(intervalId);
                                if (callback) {
                                    callback(err_1, null);
                                }
                                else {
                                    reject(err_1);
                                }
                                return [3 /*break*/, 6];
                            case 6: return [2 /*return*/];
                        }
                    });
                }); }, 10 * 1000);
                return [2 /*return*/, callback ? null : new Promise(function (_resolve, _reject) {
                        resolve = _resolve;
                        reject = _reject;
                    })];
            });
        });
    }
    function _runTestSuite(suite, paramsOrCallback, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var overrideParameters, suiteId, additionalDescription, labels, failGracefully, pool;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                overrideParameters = paramsOrCallback && typeof paramsOrCallback !== 'function' ? paramsOrCallback : {};
                suiteId = suite.id;
                additionalDescription = suite.options && suite.options.additionalDescription;
                labels = suite.options && suite.options.labels && utils_1.filterLabels(suite.options.labels);
                failGracefully = suite.options && suite.options.failGracefully;
                pool = suite.options && suite.options.pool;
                return [2 /*return*/, wrap(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                        var _a, testSuiteRunId, err;
                        return tslib_1.__generator(this, function (_b) {
                            switch (_b.label) {
                                case 0: return [4 /*yield*/, superagent.post(testSuitesAPI + "/" + suiteId + "/run" + (failGracefully ? '?failGracefully=true' : ''))
                                        .send({ overrideParameters: overrideParameters, additionalDescription: additionalDescription, labels: labels, pool: pool })
                                        .auth(token, '')];
                                case 1:
                                    _a = (_b.sent()).body, testSuiteRunId = _a.testSuiteRunId, err = _a.err;
                                    if (err || !testSuiteRunId) {
                                        console.error(err ? JSON.stringify(err) : "The server encountered an error while handling the request");
                                        return [2 /*return*/];
                                    }
                                    return [2 /*return*/, { id: testSuiteRunId, type: Loadmill.TYPES.SUITE }];
                            }
                        });
                    }); }, callback || paramsOrCallback)];
            });
        });
    }
    function _runTestPlan(testPlan, params) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var testPlanId, overrideParameters, labels, additionalDescription, pool, parallel, _a, testPlanRunId, err;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        testPlanId = testPlan.id;
                        overrideParameters = params || {};
                        labels = testPlan.options && testPlan.options.labels && utils_1.filterLabels(testPlan.options.labels);
                        additionalDescription = testPlan.options && testPlan.options.additionalDescription;
                        pool = testPlan.options && testPlan.options.pool;
                        parallel = testPlan.options && testPlan.options.parallel;
                        return [4 /*yield*/, superagent.post(testPlansAPI + "/" + testPlanId + "/run")
                                .send({ overrideParameters: overrideParameters, additionalDescription: additionalDescription, labels: labels, pool: pool, parallel: parallel })
                                .auth(token, '')];
                    case 1:
                        _a = (_b.sent()).body, testPlanRunId = _a.testPlanRunId, err = _a.err;
                        if (err || !testPlanRunId) {
                            console.error(err ? JSON.stringify(err) : "The server encountered an error while handling the request");
                            return [2 /*return*/];
                        }
                        return [2 /*return*/, { id: testPlanRunId, type: Loadmill.TYPES.TEST_PLAN }];
                }
            });
        });
    }
    function _junitReport(testResult, path) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, reporter_1.junitReport(testResult, token, path)];
            });
        });
    }
    function _mochawesomeReport(testResult, path) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, reporter_1.mochawesomeReport(testResult, token, path)];
            });
        });
    }
    return {
        run: function (config, paramsOrCallback, callback) {
            var _this = this;
            return wrap(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var testId;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            config = toConfig(config, paramsOrCallback);
                            return [4 /*yield*/, superagent.post(testingServer + "/api/tests")
                                    .send(config)
                                    .auth(token, '')];
                        case 1:
                            testId = (_a.sent()).body.testId;
                            return [4 /*yield*/, superagent.put(testingServer + "/api/tests/" + testId + "/load")
                                    .auth(token, '')];
                        case 2:
                            _a.sent();
                            return [2 /*return*/, testId];
                    }
                });
            }); }, callback || paramsOrCallback);
        },
        runFolder: function (folderPath, paramsOrCallback, callback) {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var listOfFiles;
                return tslib_1.__generator(this, function (_a) {
                    listOfFiles = utils_1.getJSONFilesInFolderRecursively(folderPath);
                    if (listOfFiles.length === 0) {
                        console.log("No Loadmill test files were found at " + folderPath + " - exiting...");
                    }
                    return [2 /*return*/, _runFolderSync(listOfFiles, this.run, paramsOrCallback, callback)];
                });
            });
        },
        wait: function (testDefOrId, callback) {
            return _wait(testDefOrId, callback);
        },
        runTestSuite: function (suite, paramsOrCallback, callback) {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    return [2 /*return*/, _runTestSuite(suite, paramsOrCallback, callback)];
                });
            });
        },
        runTestPlan: function (testPlan, params) {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    return [2 /*return*/, _runTestPlan(testPlan, params)];
                });
            });
        },
        junitReport: function (testResult, path) {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    return [2 /*return*/, _junitReport(testResult, path)];
                });
            });
        },
        mochawesomeReport: function (testResult, path) {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    return [2 /*return*/, _mochawesomeReport(testResult, path)];
                });
            });
        }
    };
}
var isTestPassed = function (body, type) {
    switch (type) {
        case Loadmill.TYPES.SUITE:
        case Loadmill.TYPES.TEST_PLAN:
            return body.status === "PASSED";
        default: //load
            return body.result === 'done';
    }
};
function isTestInFinalState(body) {
    var trialResult = body.trialResult, result = body.result, status = body.status;
    return ((result || trialResult === false) || // load tests
        (status && status !== "RUNNING") // test suites or test plan
    );
}
function getTestAPIUrl(_a, server) {
    var id = _a.id, type = _a.type;
    var prefix = server + "/api";
    switch (type) {
        case Loadmill.TYPES.SUITE:
            return prefix + "/test-suites-runs/" + id;
        case Loadmill.TYPES.TEST_PLAN:
            return prefix + "/test-plans-runs/" + id;
        default: //load
            return prefix + "/tests/" + id;
    }
}
function getTestWebUrl(_a, server) {
    var id = _a.id, type = _a.type;
    var prefix = server + "/app";
    switch (type) {
        case Loadmill.TYPES.SUITE:
            return prefix + "/api-tests/test-suite-runs/" + id;
        case Loadmill.TYPES.TEST_PLAN:
            return prefix + "/api-tests/test-plan-runs/" + id;
        default: //load
            return prefix + "/test/" + id;
    }
}
function reductFlowRunsData(flowRuns) {
    if (flowRuns) {
        return flowRuns.map(function (f) { return ({
            id: f.id,
            description: f.description,
            status: f.status
        }); });
    }
}
function reductTestSuitesRuns(suitesRuns, testingServer) {
    if (suitesRuns) {
        return suitesRuns.map(function (s) {
            var suiteRun = {
                id: s.id,
                type: Loadmill.TYPES.SUITE,
                description: s.description,
                status: s.status,
                url: getTestWebUrl({ id: s.id, type: Loadmill.TYPES.SUITE }, testingServer),
                passed: s.status === "PASSED",
                startTime: s.startTime,
                endTime: s.endTime
            };
            if (Array.isArray(s.testSuiteFlowRuns)) {
                suiteRun.flowRuns = s.testSuiteFlowRuns.map(function (fr) { return ({
                    id: fr.id,
                    status: fr.status,
                    description: fr.description
                }); });
            }
            return suiteRun;
        });
    }
}
function wrap(asyncFunction, paramsOrCallback) {
    var promise = asyncFunction();
    if (typeof paramsOrCallback === 'function') {
        promise.then(function (res) { return paramsOrCallback(null, res); })["catch"](function (err) { return paramsOrCallback(err, null); });
    }
    else {
        return promise;
    }
}
function toConfig(config, paramsOrCallback) {
    if (typeof config === 'string') {
        var text = fs.readFileSync(config).toString();
        config = JSON.parse(text);
    }
    if (typeof paramsOrCallback === 'object' && paramsOrCallback != null) {
        var parameters = config.parameters;
        if (!parameters) {
            config.parameters = paramsOrCallback;
        }
        else if (typeof parameters.push === 'function') {
            parameters.push(paramsOrCallback);
        }
        else {
            config.parameters = [parameters, paramsOrCallback];
        }
    }
    return config;
}
(function (Loadmill) {
    var TYPES;
    (function (TYPES) {
        TYPES["LOAD"] = "load";
        TYPES["SUITE"] = "test-suite";
        TYPES["TEST_PLAN"] = "test-plan";
    })(TYPES = Loadmill.TYPES || (Loadmill.TYPES = {}));
    ;
})(Loadmill || (Loadmill = {}));
module.exports = Loadmill;
